File: src/Button.cpp
#include "Button.hpp"
#include "Font.hpp"
#include "Window.hpp"
#include "Timer.hpp"

Button::Button(std::shared_ptr<Texture> normalTexture,std::shared_ptr<Texture> hoverTexture, std::shared_ptr<Texture> pressedTexture, int x, int y, int width, int height) : ElementGUI() {
	
	_rect.left = x;
	_rect.top = y;
	_rect.width = width;
	_rect.height = height;
	
	_normalTexture = normalTexture;
	_hoverTexture = hoverTexture;
	_pressedTexture = pressedTexture;
	
	_state = ButtonState::Idle;
	
	_onclick_func =  { };
}

Button::~Button() { }


void Button::unclick() {
        _state = ButtonState::Idle;
}

void Button::hover() {
        _state = ButtonState::Hover;
}

void Button::click() {
        _state = ButtonState::Pressed;
        _clickTime = currentTime;
}

void Button::handleEvents(sf::Event& event) {
	sf::Vector2f touch = window->mapPixelToCoords(sf::Vector2i(event.touch.x, event.touch.y));
	if(_rect.contains(touch.x, touch.y)) {
		if(event.type == sf::Event::TouchBegan) {
			ElementGUI_pressed = this->shared_from_this();
		}
		else if(event.type == sf::Event::TouchEnded) {
			if (ElementGUI_pressed.get() == this) {
            	click();
				return;
			}  
		}
	}
	
	if(_state!=ButtonState::Pressed) {
		if(event.type == sf::Event::TouchMoved || event.type == sf::Event::TouchBegan){
			
			if(_rect.contains(touch.x, touch.y)){
				ElementGUI_hovered = this->shared_from_this();
			}else {
					
				if(ElementGUI_hovered.get()== this){
					ElementGUI_hovered = nullptr;
				}
			}
		}
	}
		
		
}

void Button::update() {
	        if (_state == ButtonState::Pressed) {
                if ((currentTime - _clickTime).asSeconds() > 0.05f) {
                        if (_onclick_func) {
                                _onclick_func();
                        }
                        
                        if(ElementGUI_pressed.get() == this)
                        	ElementGUI_pressed = nullptr;
                        unclick();
                }
        }
        else if (ElementGUI_hovered.get() == this) {
                hover();
        }
        else
                unclick();
}

void Button::draw() {
	
	sf::Texture* tex = nullptr;
	switch (_state) {
	    case ButtonState::Idle:
	        tex = &_normalTexture->_texture;
	        break;
	    case ButtonState::Hover:
	        tex = &_hoverTexture->_texture;
	        break;
	    case ButtonState::Pressed:
	        tex = &_pressedTexture->_texture;
	        break;
	}

	sf::Sprite spr(*tex);
	spr.setTextureRect(sf::IntRect(0, 0, _rect.width, _rect.height));
	spr.setPosition(_rect.left, _rect.top);
	window->draw(spr);
}

ButtonWithText::ButtonWithText(std::wstring text, std::shared_ptr<Texture> normalTexture,std::shared_ptr<Texture> hoverTexture, std::shared_ptr<Texture> pressedTexture, int x, int y, int width, int height) : Button(normalTexture, hoverTexture, pressedTexture, x, y, width, height) {
	
	int textSize = 40;
	
	_text = std::make_unique<sf::Text>(text, basicFont, textSize);
	float xx = _rect.left + (_rect.width - _text->getGlobalBounds().width)/2.f;
	float yy = _rect.top + (_rect.height - basicFont.getLineSpacing(textSize))/2.f;
	_text->setPosition(xx, yy);
}

ButtonWithText::~ButtonWithText() { }

void ButtonWithText::handleEvents(sf::Event& event) {
	Button::handleEvents(event);
}

void ButtonWithText::update() {
	Button::update();
}

void ButtonWithText::draw() {
	Button::draw();
	window->draw(*_text);
}

File: core/src/main/java/com/tbane/mysticgems/GUI/Button.java
package com.tbane.mysticgems.GUI;

import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.math.Rectangle;

import com.tbane.mysticgems.MyInput.MyInput;
import com.tbane.mysticgems.Time;
import com.tbane.mysticgems.Renderer;

public class Button extends ElementGUI {

    public Rectangle _rect;
    private final Texture _normalTexture, _hoverTexture, _pressedTexture;
    public enum states { Idle, Hover, Pressed }
    public states _state;
    float _clickTime;
    public Runnable onclick_func;

    public Button(Texture normalTexture, Texture hoverTexture, Texture pressedTexture, int x, int y, int width, int height) {
        super();
        _rect = new Rectangle(x, y, width, height);
        _normalTexture = normalTexture;
        _hoverTexture = hoverTexture;
        _pressedTexture = pressedTexture;


        _state = states.Idle;
        _clickTime = 0.0f;

        onclick_func = () -> {};
    }
    private void unclick() {
        _state = states.Idle;
    }

    private void hover() {
        _state = states.Hover;
    }

    private void click() {
        _state = states.Pressed;
        _clickTime = Time.currentTime;
    }

    @Override
    public void handleEvents() {



        if(_rect.contains(MyInput.processor.getTouchX(), MyInput.processor.getTouchY())) {
            if(MyInput.processor.isTouchDown()) {
                GUI.ElementGUI_pressed = this;
            }
		else if(MyInput.processor.isTouchUp()) {
                if (GUI.ElementGUI_pressed == this) {
                    click();
                    return;
                }
            }
        }

        if(_state!= states.Pressed) {
            if(MyInput.processor.isTouchMoved() || MyInput.processor.isTouchDown()){

                if(_rect.contains(MyInput.processor.getTouchPosition())){
                    GUI.ElementGUI_hovered = this;
                }else {

                    if(GUI.ElementGUI_hovered == this){
                        GUI.ElementGUI_hovered = null;
                    }
                }
            }
        }

    }

    @Override
    public void update() {
        if (_state == states.Pressed) {
            if ((Time.currentTime - _clickTime) > 0.05f) {
                onclick_func.run();

                if(GUI.ElementGUI_pressed == this)
                    GUI.ElementGUI_pressed = null;
                unclick();
            }
        }
        else if (GUI.ElementGUI_hovered == this) {
            hover();
        }
        else
            unclick();
    }

    @Override
    public void draw() {
        Texture tex;

        switch (_state) {
            case Hover:
                tex = _hoverTexture;
                break;
            case Pressed:
                tex = _pressedTexture;
                break;
            default:
                tex = _normalTexture;
                break;
        }

        Sprite sprite = new Sprite(tex);
        sprite.setBounds(_rect.x, _rect.y, _rect.width, _rect.height);
        sprite.draw(Renderer.batch);
    }
}
